roachtest: scrape prometheus metrics on demand

otan opened the pull request:
This commit adds a prometheus scraper on TPC-C workloads and scrapes
their metrics. This is configured by setting it in the tpccOptions
setup. It is currently set up on the multi-region chaos tests.

At the end of the run, we take a snapshot of the metrics of the workload
and store it as an artifact, which can be used later.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/66657)
<!-- Reviewable:end -->

stevendanna commented on pkg/cmd/roachtest/prometheus.go:
```diff
@@ -0,0 +1,166 @@
+		"install prometheus",
+		fmt.Sprintf(
+			`rm -rf /tmp/prometheus && mkdir /tmp/prometheus && cd /tmp/prometheus &&
+			wget -O prometheus.tar.gz https://github.com/prometheus/prometheus/releases/download/v2.27.1/prometheus-2.27.1.linux-amd64.tar.gz &&
```
[nit] It doesn't matter, but we use `curl` nearly everywhere else for downloading things.  If you do move to curl, you'll want the `--fail` flag most likely.

stevendanna replied:
This currently won't work for local runs for non-linux workstations.  Are there differences other than the binary that we are using?  If so, perhaps we could add a TODO comment with some of the differences and output a better error message if you are trying to run on a local cluster and your `runtime.GOOS` is not linux.  If not, perhaps a small function that returns the right download URL for the given platform so that this would work on local clusters?

stevendanna commented on pkg/cmd/roachtest/prometheus.go:
```diff
@@ -0,0 +1,166 @@
+			ctx,
+			pm.prometheusNode,
+			`cd /tmp/prometheus &&
+			./prometheus --config.file=prometheus.yml --storage.tsdb.path=data/ --web.enable-admin-api`,
```
🤔 We've recently converted our cockroach startup to use systemd-run. I wonder if it would be worth it to do that here as well.

tbg replied:
Yes, really good idea. We can `PutString` the entry point script (if we even need one) to keep this file clean.
Ultimately, we want everything to be launched on these clusters through systemd. While that's a pipedream as of yet (due to the large numbers of one-off commands) it is where we should be heading.

tbg replied:
The existing systemd code is here: https://github.com/cockroachdb/cockroach/blob/98a75bb3b9ad10c27336fbe76c92489b4cdbf6f2/pkg/cmd/roachprod/install/cockroach.go#L378-L458

The self-installing re-entrant bash script pattern is certainly not the part to copy here. Just upload a script (if needed), install the job, start the job.

stevendanna commented on pkg/cmd/roachtest/prometheus.go:
```diff
@@ -0,0 +1,166 @@
+	if err := c.RunE(
+		ctx,
+		pm.prometheusNode,
+		`pgrep prometheus | xargs kill -2`,
```
[nit] Since we are depending on `pgrep` anyway, we could probably use `pkill` here.

stevendanna commented on pkg/cmd/roachtest/prometheus.go:
```diff
@@ -0,0 +1,166 @@
+			`rm -rf /tmp/prometheus && mkdir /tmp/prometheus && cd /tmp/prometheus &&
+			wget -O prometheus.tar.gz https://github.com/prometheus/prometheus/releases/download/v2.27.1/prometheus-2.27.1.linux-amd64.tar.gz &&
+			tar xvf prometheus.tar.gz --strip-components=1 &&
+			echo "%s" > prometheus.yml
```
Might be a little nicer to use PutString for this configuration to avoid confusing errors if we ever need quoted strings in our yaml configuration.

tbg replied:
Was going to say the same, `PutString` is nicer here. Heck, I'd even start bundling these scripts with go:embed but that's not something I'd force upon you here, but if you're so inclined I'd be happy to see it.

tbg commented on pkg/cmd/roachtest/cluster_interface_mock_test.go:
```diff
@@ -0,0 +1,759 @@
+// Code generated by MockGen. DO NOT EDIT.
```
Interesting that GH doesn't detect this as generated code.

tbg commented on pkg/cmd/roachtest/cluster_interface.go:
```diff
@@ -7,8 +7,11 @@
+
 package main
 
+//go:generate mockgen -package=main -destination=cluster_interface_mock_test.go -source=cluster_interface.go . Cluster
```
Please add a comment on `Cluster` that mentions that `go generate ./pkg/cmd/roachtest/...` is in order when changing the interface.

tbg commented on pkg/cmd/roachtest/prometheus_test.go:
```diff
@@ -0,0 +1,88 @@
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
```
❤️

tbg commented on pkg/cmd/roachtest/prometheus.go:
```diff
@@ -0,0 +1,166 @@
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package main
```
Curious what you think about putting this in its own package. I would do it without blinking - prometheus is an "addon" that can be installed onto a cluster, and shouldn't need access to any internals.

I realize this isn't easy right now (`Cluster` sits in `main`, `main` needs to import the new package, the new package needs to import main...) but once `Cluster` is in a leaf pkg, it should work.

Not saying this needs to be done now, just please don't refer to anything from `main` in here so we can do it later (I think you're already cognizant of that).

stevendanna commented on pkg/cmd/roachtest/tpcc.go:
```diff
@@ -134,6 +136,41 @@ func setupTPCC(
 func runTPCC(ctx context.Context, t *test, c Cluster, opts tpccOptions) {
+	if cfg := opts.PrometheusConfig; cfg != nil {
+		if c.isLocal() {
+			t.Status("skipping test as prometheus is needed, but prometheus does not yet work locally")
```
Perhaps t.Skip here?

stevendanna commented on pkg/cmd/roachtest/prometheus/prometheus.go:
```diff
@@ -0,0 +1,209 @@
+
+// MakeInsecureCockroachScrapeConfig creates scrape configs for the given
+// cockroach nodes. All nodes are assumed to be insecure and running on
+// port 26258.
```
I believe there is a way to get the port via roachprod if this function were to take a *cluser as well.  But, I think that will com naturally if we decide to follow-up with local-mode support.

stevendanna reviewed (approved):
Left a minor comment or two, but I leave them to your discretion.

Thanks for addressing the previous comments

otan committed:
roachtest: scrape prometheus metrics on demand

This commit adds a prometheus scraper on TPC-C workloads and scrapes
their metrics. This is configured by setting it in the tpccOptions
setup. It is currently set up on the multi-region chaos tests.

At the end of the run, we take a snapshot of the metrics of the workload
and store it as an artifact, which can be used later.

Release note: None

otan commented:
thanks!

bors r=stevendanna
